packages/prover/src/proof_provider/proof_provider.ts

```diff
@@ -0,0 +1,200 @@
+      finalized: this.lightClient.getFinalized().beacon.slot,
+      status: this.lightClient.getStatus(),
+    };
+  }
+
+  getExecutionPayload(blockNumber: number | string | "finalized" | "latest"): allForks.ExecutionPayload {
+    assertLightClient(this.lightClient);
+
+    if (typeof blockNumber === "string" && blockNumber === "finalized") {
+      const payload = this.finalizedPayloads.last;
+      if (!payload) throw new Error("No finalized payload");
+      return payload;
+    }
+
+    if (typeof blockNumber === "string" && blockNumber === "latest") {
+      const payload = this.payloads.last;
+      if (!payload) throw new Error("No latest payload");
+      return payload;
+    }
+
+    if (typeof blockNumber === "string" && blockNumber.startsWith("0x")) {
+      const payload = this.payloads.get(blockNumber);
+      if (!payload) throw new Error(`No payload for blockNumber ${blockNumber}`);
+      return payload;
+    }
+
+    if (typeof blockNumber === "number") {
+      const payload = this.payloads.get(numberToHex(blockNumber));
+      if (!payload) throw new Error(`No payload for blockNumber ${blockNumber}`);
+      return payload;
+    }
+
+    throw new Error(`Invalid blockNumber "${blockNumber}"`);
+  }
+
+  async update(update: allForks.LightClientHeader, finalized = false): Promise<void> {
+    const fork = this.options.config.getForkName(update.beacon.slot);
+
+    if (!isForkExecution(fork)) {
+      return;
+    }
+
+    if (isForkExecution(fork) && !("execution" in update)) {
+      throw new Error("Execution payload is required for execution fork");
+    }
+
+    const newPayload = (update as capella.LightClientHeader).execution;
+    const blockNumber = numberToHex(newPayload.blockNumber);
+    const blockSlot = update.beacon.slot;
+    const existingPayload = this.payloads.get(blockNumber);
+
+    if (existingPayload && existingPayload.blockHash === newPayload.blockHash) {
+      if (finalized) {
+        this.finalizedPayloads.set(blockNumber, existingPayload);
+      }
+
+      // We payload have the payload for this block
+      return;
+    }
+
+    if (existingPayload && existingPayload.blockHash !== newPayload.blockHash) {
+      // TODO: Need to decide either
+      // 1. throw an error
+      // 2. or just ignore it
+      // 3. or update the blockHash
+    }
+
+    const newPayloads = await getExecutionPayloads(this.options.api, blockSlot, blockSlot);
+    this.payloads.set(blockNumber, newPayloads[blockSlot]);
+    this.slotsMap[blockSlot] = blockNumber;
+    if (finalized) {
+      this.finalizedPayloads.set(blockNumber, newPayloads[blockSlot]);
+    }
+  }
+
+  private registerEvents(): void {
+    assertLightClient(this.lightClient);
+
+    this.options.signal.addEventListener("abort", () => {
+      this.lightClient?.stop();
+    });
+
+    this.lightClient.emitter.on(LightclientEvent.lightClientFinalityUpdate, async (data) => {
+      await this.update(data, true).catch((e) => {
```

Comment:
aren't you `typing` arg of lightclient header to this.update, data you should get here is of type update

Also it would be better to emit `latest` and `finalized` events from lightclient as thats what you mean to process here i think

Reply:
I didn't get that comment. Please explain it a bit.

Reply:
https://github.com/ChainSafe/lodestar/pull/5222/files#diff-229bc1a2e4f16b99b96ae3bd25f0429ff996cf967786ac6235f53f8f4a5b1e2cR142 
not sure the types match

Reply:
Yes the type matches to the `lightclient` events else TS would not have let pass it. Remember it's not the direct event handler, rather a function called from within event handler without any type casting.